sorting/sorting.go

```diff
@@ -0,0 +1,123 @@
+package sorting
+
+import (
+	"bufio"
+	"errors"
+	"log"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// a defined type to simulate a tuple with three elements as sepecified in the associated function
+type Tuple []int
+
+func (tuple *Tuple) appendOne(number int) error {
+	/* This function appends one number to the defined type Tuple if it's not already full (with 3 numbers)
+	otherwise it'll return an error that will be logged using the function HandleErrors */
+	if len(*tuple) > 3 {
+		return errors.New("cannot add anymore numbers")
+	}
+	*tuple = append(*tuple, number)
+	return nil
+}
+
+func ReadingTuplesFromFile(fileName string) []Tuple {
+	/* Reads the input from the given files specified in the argument fileName, checks for errors in the process of
+	opening the file, the reads it line by line, and uses a helper function to cast the strings into ints for the tuple
+	then if the tuple consists of 3 elements (the main condition), it adds it to the list */
+	var listOfTuples []Tuple
+	var file *os.File
+	var err error
+	file, err = os.Open(fileName)
+	handleErrors(err)
+	defer file.Close()
+	var scanner *bufio.Scanner = bufio.NewScanner(file)
+	for scanner.Scan() {
+		var line string = scanner.Text()
+		var tuple Tuple = stringToInt(line)
+		if len(tuple) == 3 {
+			listOfTuples = append(listOfTuples, tuple)
+		}
+	}
+	return listOfTuples
+}
+
+func handleErrors(err error) {
+	// Logging the error passed as an argument to it, checks whether or not everything's ok
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func stringToInt(line string) Tuple {
+	/* a helper function that checks for a specific expression to make sure that the condition is met
+	then casts the string values into integer values to be stored in a tuple */
+	var regExp *regexp.Regexp
+	var err error
+	regExp, err = regexp.Compile(`^\d,\s?\d,\s?\d$`)
+	handleErrors(err)
+	var tuple Tuple
+	if regExp.MatchString(line) {
+		var stringParts []string = strings.Split(line, ",")
+		for _, value := range stringParts {
+			value = strings.TrimSpace(value)
+			var intValue int
+			intValue, err = strconv.Atoi(value)
+			handleErrors(err)
+			tuple.appendOne(intValue)
+		}
+	}
+	return tuple
+}
+
+func SortList(list []Tuple, column int) []Tuple {
+	/* Sorts the list of tuples based on the columns specified to be sorted by
+	the column values and the tuples are bound with a map, with respect to the condition of non-unique keys */
+	var mapping map[int]Tuple = make(map[int]Tuple)
+	for _, segment := range list {
+		if _, ok := mapping[segment[column]]; ok {
```

Comment:
Few comments on the logic in this function:
1. Space complexity: it duplicates the entire input set to a secondary hash-map. It duplicates again one column to another array. It duplicates the entire input set to the final list. That's expensive.
2. Deep in the code, we see here a piece of code that restricts us from sorting non-unique columns. Considering that there's no good reason to do it, we should not. If there's good reason, that should be enforced by the API not by the sort function.
3. What you actually want here is a map from `value-in-sort-column -> index-of-tuple-in-input-list` right? This reduces the size and handle duplicates.